optimum/habana/transformers/generation/utils.py

```diff
@@ -1239,8 +1243,426 @@ def contrastive_search(
+        eos_token_id = eos_token_id if eos_token_id is not None else self.generation_config.eos_token_id
+        sequential = sequential if sequential is not None else self.generation_config.low_memory
+        if isinstance(eos_token_id, int):
+            eos_token_id = [eos_token_id]
+        eos_token_id_tensor = torch.tensor(eos_token_id).to(input_ids.device) if eos_token_id is not None else None
+        output_scores = output_scores if output_scores is not None else self.generation_config.output_scores
+        output_logits = output_logits if output_logits is not None else self.generation_config.output_logits
+        output_attentions = (
+            output_attentions if output_attentions is not None else self.generation_config.output_attentions
+        )
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.generation_config.output_hidden_states
+        )
+        return_dict_in_generate = (
+            return_dict_in_generate
+            if return_dict_in_generate is not None
+            else self.generation_config.return_dict_in_generate
+        )
+
+        # init attention / hidden states / scores tuples
+        raw_logits = () if (return_dict_in_generate and output_logits) else None
+        scores = () if (return_dict_in_generate and output_scores) else None
+        decoder_attentions = () if (return_dict_in_generate and output_attentions) else None
+        cross_attentions = () if (return_dict_in_generate and output_attentions) else None
+        decoder_hidden_states = () if (return_dict_in_generate and output_hidden_states) else None
+
+        # if model is an encoder-decoder, retrieve encoder attention weights and hidden states
+        if return_dict_in_generate and self.config.is_encoder_decoder:
+            encoder_attentions = model_kwargs["encoder_outputs"].get("attentions") if output_attentions else None
+            encoder_hidden_states = (
+                model_kwargs["encoder_outputs"].get("hidden_states") if output_hidden_states else None
+            )
+
+        # keep track of which sequences are already finished
+        unfinished_sequences = torch.ones(input_ids.shape[0], dtype=torch.long, device=input_ids.device)
+
+        hb_profer = HabanaProfile(warmup=profiling_warmup_steps, active=profiling_steps)
+        hb_profer.start()
+        this_peer_finished = False  # used by synced_gpus only
+        batch_size, cur_len = input_ids.shape
+        bucket_size = model_kwargs.get("bucket_size", -1)
+        bucket_internal = model_kwargs.get("bucket_internal", None)
+        reduce_recompile = model_kwargs.get("reduce_recompile", False)
+
+        prompt_len = cur_len
+        if not bucket_internal:
+            if bucket_size >= 0:
+                inc = iter(incrementor(bucket_size, prompt_len))
+            if bucket_size > 0:  # TODO: Why is this needed given above statement?
```

Comment:
@regisss  @jiminha I copied this code from `greedy_search`. I would like clarification on why it is needed. I added the todo comment to confirm whether to keep the statement. 

Should I also modify `is_greedy_or_beam_and_bucket` to include contrastive_search? https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/generation/utils.py#L587

Reply:
When I originally added the assert in greedy, I was seeing that position_ids was not present when I had bucketing. I think I had that assert just in case somehow position ids was passed, and I wanted that assert to be hit, so that we could study what was teh use case and deal with it appropriately.



Yes u can modify is_greedy_or_beam_and_bucket

Btw what are the commands you are testing?

Reply:
@ssarkar2 Thanks for the clarification. I will keep the assertion so that we can study it if we encounter it. I'll add contrastive search to is_greedy_or_beam_and_bucket criteria. 

I have included the pytests that are fixed by this PR in the PR description. They do not test the bucketing feature. Is there a test or script I can run to test it?

Reply:
@ssarkar2  I added the contrastive search arguments to the `run_generation.py` and `README`. 

Here are sample outputs.

Greedy search:
`python run_generation.py --model_name_or_path meta-llama/Llama-2-7b-hf --use_hpu_graphs --use_kv_cache --batch_size 1 --bf16 --max_new_tokens 512 --prompt "DeepMind Company is"`

'DeepMind Company is a British artificial intelligence company. It was founded in 2010 by Demis Hassabis, Mustafa Suleyman, and Shane Legg. The company is based in London, England.\nDeepMind is a subsidiary of Google. In 2014, Google acquired DeepMind for $400 million.\nDeepMind is best known for its work in artificial intelligence. The company has developed a number of AI systems, including AlphaGo, which defeated a professional Go player in 2016.\nDeepMind is also working on a number of other projects, including a system that can help doctors diagnose diseases and a system that can help people learn new skills.\nDeepMind is a British artificial intelligence company that was founded in 2010. The company is best known for its work in developing artificial intelligence systems that can play games and solve problems. DeepMind has also developed a number of other AI systems, including a system that can help doctors diagnose diseases and a system that can help people learn new skills.\nDeepMind is a British artificial intelligence company that was founded in 2010. The company is best known for its work in developing artificial intelligence systems that can play games and solve problems. DeepMind has also developed a number of other AI systems, including a system that can help doctors diagnose diseases and a system that can help people learn new skills. DeepMind is a subsidiary of Google.\nDeepMind is a British artificial intelligence company that was founded in 2010. The company is best known for its work in developing artificial intelligence systems that can play games and solve problems. DeepMind has also developed a number of other AI systems, including a system that can help doctors diagnose diseases and a system that can help people learn new skills. DeepMind is a subsidiary of Google. DeepMind is a British artificial intelligence company that was founded in 2010. The company is best known for its work in developing artificial intelligence systems that can play games and solve problems. DeepMind has also developed a number of other AI systems, including a system that can help doctors diagnose diseases and a system that can help people learn new skills. DeepMind is a subsidiary of Google.\nDeepMind is a

`python run_generation.py --model_name_or_path meta-llama/Llama-2-7b-hf --use_hpu_graphs --use_kv_cache --batch_size 1 --bf16 --max_new_tokens 512 --prompt "DeepMind Company is" --top_k 4 --penalty_alpha 0.6`

DeepMind Company is a British artificial intelligence (AI) company.\nDeepMind Technologies Limited is a subsidiary of Alphabet Inc. that specializes in AI research and its applications. The company's headquarters are in London, United Kingdom.\nIn 2010, Demis Hassabis, Shane Legg, Mustafa Suleyman, and Wojciech Zaremba founded DeepMind, an Artificial Intelligence company. They were joined by Peter Wurman, Othman Larouz, James Kvamqvist, Max Jaderberg, Oriol Vinyals, Quoc Le Nguyen, Ilya Sutskevichen, Sven Scheurell, David Reeves-Crookshank, Tom Schaeffer, Richard Suttner, and other researchers.\nOn 14 May 2014, Google announced that it had acquired DeepMind for $400 million in cash and $400 million in Google equity. The acquisition was completed on 1 August 2014.\nIn 2015, DeepMind's researchers published a paper in Nature that described an algorithm that could learn to play Atari games without being programmed to do so. This was the first time that a computer had been able to master a video game by learning from experience, rather than relying on pre-programmed rules.\nIn 2016, DeepMind's researchers published a paper in Nature Neuroscience that described a neural network that could predict the movements of people with Parkinson's disease with 96% accuracy. This was the first time that a machine had been able to accurately predict the movements of people with a neurological disorder.\nIn 2017, DeepMind's researchers published a paper in Nature Communications that described a neural network that could generate realistic-looking images of human faces. This was the first time that a machine had been able to generate convincing images of human faces.\nIn 2018, DeepMind's researchers published a paper in Nature Machine Intelligence that described a neural network that could solve Rubik's Cube puzzles in 1.2 seconds, which is faster than any human can do it.